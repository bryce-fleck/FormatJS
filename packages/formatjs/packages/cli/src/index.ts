export {default as extract, ExtractCLIOptions} from './extract';
