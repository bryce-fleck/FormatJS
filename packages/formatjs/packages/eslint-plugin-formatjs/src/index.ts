import enforceDescription from './rules/enforce-description';
import noCamelCase from './rules/no-camel-case';
import noEmoji from './rules/no-emoji';
import noMultiplePlurals from './rules/no-multiple-plurals';
import noOffset from './rules/no-offset';
import blacklistElements from './rules/blacklist-elements';
const plugin = {
  rules: {
    'enforce-description': enforceDescription,
    'no-camel-case': noCamelCase,
    'no-emoji': noEmoji,
    'no-multiple-plurals': noMultiplePlurals,
    'no-offset': noOffset,
    'blacklist-elements': blacklistElements,
  },
};

export type Plugin = typeof plugin;

module.exports = plugin;
