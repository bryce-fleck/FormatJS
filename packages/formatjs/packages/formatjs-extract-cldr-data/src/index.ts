/*
 * Copyright 2015, Yahoo Inc.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */
import expandLocales, {LocaleEntry} from './expand-locales';
import extractRelativeFields, {FieldData} from './extract-relative';
import extractUnits, {Units, UnitData} from './extract-units';
import {getAllLocales} from './locales';
import {Locale} from './types';

export interface Opts {
  locales?: string[];
  relativeFields?: boolean;
}

export {process as processAliases} from './process-aliases';

export function extractAllRelativeFields(options: Opts = {}) {
  // Default to all CLDR locales if none have been provided.
  const locales = options.locales || getAllLocales();

  // Each type of data has the structure: `{"<locale>": {"<key>": <value>}}`,
  // which is well suited for merging into a single object per locale. This
  // performs that deep merge and returns the aggregated result.
  return mergeData<Record<string, FieldData>>(
    expandLocales(locales),
    extractRelativeFields(locales)
  );
}

export function extractAllUnits(options: Opts = {}) {
  // Default to all CLDR locales if none have been provided.
  const locales = options.locales || getAllLocales();
  const allUnitsData = extractUnits(locales);
  return Object.keys(allUnitsData).reduce(
    (
      all: Record<string, Record<Locale, LocaleEntry & {fields?: UnitData}>>,
      unit
    ) => {
      all[unit] = mergeData<UnitData>(
        expandLocales(locales),
        allUnitsData[unit]
      );
      return all;
    },
    {}
  );
}

function mergeData<D extends Record<string, any>>(
  entries: Record<Locale, LocaleEntry> = {},
  fields: Record<Locale, {fields: D}> = {}
): Record<Locale, LocaleEntry & {fields?: D}> {
  const data: Record<Locale, LocaleEntry & {fields?: D}> = {};
  Object.keys(entries).forEach(locale => {
    data[locale] = {...entries[locale]};
  });
  Object.keys(fields).forEach(locale => {
    if (!data[locale]) {
      data[locale] = {locale};
    }
    data[locale] = {...data[locale], ...fields[locale]};
  });
  return data;
}
