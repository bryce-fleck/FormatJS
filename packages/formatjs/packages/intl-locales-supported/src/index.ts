/*
 * Copyright 2015, Yahoo Inc.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */
const DEFAULT_CONSTRUCTORS = [
  Intl.DateTimeFormat,
  Intl.NumberFormat,
  Intl.PluralRules
];
export default function areIntlLocalesSupported(
  locales: string | string[],
  constructorsToCheck = DEFAULT_CONSTRUCTORS
): boolean {
  if (typeof Intl === 'undefined') {
    return false;
  }

  if (!locales) {
    throw new Error('locales must be supplied.');
  }

  if (!Array.isArray(locales)) {
    locales = [locales];
  }

  const intlConstructors = constructorsToCheck.filter(Boolean);

  if (intlConstructors.length === 0) {
    return false;
  }

  return intlConstructors.every(
    intlConstructor =>
      intlConstructor.supportedLocalesOf(locales).length === locales.length
  );
}
