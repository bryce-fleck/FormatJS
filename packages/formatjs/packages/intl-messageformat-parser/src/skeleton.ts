/**
 * https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
 * Credit: https://github.com/caridy/intl-datetimeformat-pattern/blob/master/index.js
 * with some tweaks
 */
const DATE_TIME_REGEX = /(?:[Eec]{1,6}|G{1,5}|[Qq]{1,5}|(?:[yYur]+|U{1,5})|[ML]{1,5}|d{1,2}|D{1,3}|F{1}|[abB]{1,5}|[hkHK]{1,2}|w{1,2}|W{1}|m{1,2}|s{1,2}|[zZOvVxX]{1,4})(?=([^']*'[^']*')*[^']*$)/g;

export interface ExtendedDateTimeFormatOptions
  extends Intl.DateTimeFormatOptions {
  hourCycle?: 'h11' | 'h12' | 'h23' | 'h24';
}

/**
 * Parse Date time skeleton into Intl.DateTimeFormatOptions
 * Ref: https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
 * @public
 * @param skeleton skeleton string
 */
export function parseDateTimeSkeleton(
  skeleton: string
): ExtendedDateTimeFormatOptions {
  const result: ExtendedDateTimeFormatOptions = {};
  skeleton.replace(DATE_TIME_REGEX, match => {
    const len = match.length;
    switch (match[0]) {
      // Era
      case 'G':
        result.era = len === 4 ? 'long' : len === 5 ? 'narrow' : 'short';
        break;
      // Year
      case 'y':
      case 'Y':
      case 'u':
      case 'U':
      case 'r':
        result.year = len === 2 ? '2-digit' : 'numeric';
        break;
      // Quarter
      case 'q':
      case 'Q':
        throw new RangeError('Quarter (q/Q) is not supported');
      // Month
      case 'M':
      case 'L':
        result.month = ['numeric', '2-digit', 'short', 'long', 'narrow'][
          len - 1
        ];
        break;
      // Week
      case 'w':
      case 'W':
        throw new RangeError('Week (w/W) is not supported');
      case 'd':
        result.day = ['numeric', '2-digit'][len - 1];
        break;
      case 'D':
      case 'F':
      case 'g':
        throw new RangeError('Day (D/F/g) is not supported');
      // Weekday
      case 'E':
        result.weekday = len === 4 ? 'short' : len === 5 ? 'narrow' : 'short';
        break;
      case 'e':
        if (len < 4) {
          throw new RangeError('Weekday e..eee is not supported');
        }
        result.weekday = ['short', 'long', 'narrow', 'short'][len - 4];
        break;
      case 'c':
        if (len < 4) {
          throw new RangeError('Weekday c..ccc is not supported');
        }
        result.weekday = ['short', 'long', 'narrow', 'short'][len - 4];
        break;

      // Period
      case 'a': // AM, PM
      case 'b': // am, pm, noon, midnight
      case 'B': // flexible day periods
        result.hour12 = true;
        break;

      // Hour
      case 'h':
        result.hourCycle = 'h12';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'H':
        result.hourCycle = 'h23';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'K':
        result.hourCycle = 'h11';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'k':
        result.hourCycle = 'h24';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'j':
      case 'J':
      case 'C':
        throw new RangeError('Pattern (j/J/C) is not supported');
      // Minute
      case 'm':
        result.minute = ['numeric', '2-digit'][len - 1];
        break;
      case 's':
        result.second = ['numeric', '2-digit'][len - 1];
        break;
      case 'S':
      case 'A':
        throw new RangeError('Pattern (S/A) is not supported');
      // Zone
      case 'z': // 1..3, 4: specific non-location format
      case 'Z': // 1..3, 4, 5: The ISO8601 varios formats
      case 'O': // 1, 4: miliseconds in day short, long
      case 'v': // 1, 4: generic non-location format
      case 'V': // 1, 2, 3, 4: time zone ID or city
      case 'X': // 1, 2, 3, 4: The ISO8601 varios formats
      case 'x': // 1, 2, 3, 4: The ISO8601 varios formats
        // this polyfill only supports much, for now, we are just doing something dummy
        result.timeZoneName = len < 4 ? 'short' : 'long';
        break;
    }
    return '';
  });
  return result;
}
