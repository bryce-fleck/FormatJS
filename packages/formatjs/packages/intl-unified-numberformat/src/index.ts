export {default as UnifiedNumberFormat} from './units';
export {Unit} from './units-constants';
