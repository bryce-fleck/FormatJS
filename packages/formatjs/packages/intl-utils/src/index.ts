export {selectUnit} from './diff';
export {resolveSupportedLocales} from './locale-lookup';
