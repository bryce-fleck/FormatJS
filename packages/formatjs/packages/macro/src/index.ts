export function defineMessage<T>(msg: T): T {
  return msg;
}

export function defineMessages<T>(msgs: Record<string, T>): Record<string, T> {
  return msgs;
}

export const _ = defineMessage;
