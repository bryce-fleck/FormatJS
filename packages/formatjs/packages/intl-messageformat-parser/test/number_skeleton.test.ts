import {pegParse} from '../src/parser';
import {printAST} from '../src/printer';
import {isNumberElement, TYPE, isNumberSkeleton} from '../src/types';
import {convertNumberSkeletonToNumberFormatOptions} from '../src/skeleton';

test.each([
  `compact-short currency/GBP`,
  `@@#`,
  `currency/CAD unit-width-narrow`,
])('case: %p', skeleton => {
  const ast = pegParse(`{0, number, ::${skeleton}}`);
  expect(ast).toMatchSnapshot();
  expect(printAST(ast)).toMatchSnapshot();
});

test.each([
  'percent .00/@##',
  'percent .00/@@@',
  'percent .00/@@@@+',
  'currency/CAD .',
  'currency/GBP .0+/@@@',
  'currency/GBP .00##/@@@',
])('[convertNumberSkeletonToNumberFormatOptions] case: %p', skeleton => {
  const ast = pegParse(`{0, number, ::${skeleton}}`);
  const el = ast[0];
  if (!isNumberElement(el)) {
    throw expect(el.type).toEqual(TYPE.number);
  }
  if (!isNumberSkeleton(el.style)) {
    throw 'element style should be parsed as number skeleton';
  }
  expect(
    convertNumberSkeletonToNumberFormatOptions(el.style.tokens)
  ).toMatchSnapshot();
});
