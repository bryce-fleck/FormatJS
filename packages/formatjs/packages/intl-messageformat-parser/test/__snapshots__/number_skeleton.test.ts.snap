// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[convertNumberSkeletonToNumberFormatOptions] case: "currency/CAD ." 1`] = `
Object {
  "currency": "CAD",
  "maximumFractionDigits": 0,
  "style": "currency",
}
`;

exports[`[convertNumberSkeletonToNumberFormatOptions] case: "currency/GBP .0+/@@@" 1`] = `
Object {
  "currency": "GBP",
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 1,
  "minimumSignificantDigits": 3,
  "style": "currency",
}
`;

exports[`[convertNumberSkeletonToNumberFormatOptions] case: "currency/GBP .00##/@@@" 1`] = `
Object {
  "currency": "GBP",
  "maximumFractionDigits": 4,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 3,
  "style": "currency",
}
`;

exports[`[convertNumberSkeletonToNumberFormatOptions] case: "percent .00/@##" 1`] = `
Object {
  "maximumFractionDigits": 2,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 1,
  "style": "percent",
}
`;

exports[`[convertNumberSkeletonToNumberFormatOptions] case: "percent .00/@@@" 1`] = `
Object {
  "maximumFractionDigits": 2,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 3,
  "style": "percent",
}
`;

exports[`[convertNumberSkeletonToNumberFormatOptions] case: "percent .00/@@@@+" 1`] = `
Object {
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 4,
  "style": "percent",
}
`;

exports[`case: "@@#" 1`] = `
Array [
  Object {
    "style": Object {
      "tokens": Array [
        Object {
          "options": Array [],
          "stem": "@@#",
        },
      ],
      "type": 0,
    },
    "type": 2,
    "value": "0",
  },
]
`;

exports[`case: "@@#" 2`] = `"{0, number, ::@@#}"`;

exports[`case: "compact-short currency/GBP" 1`] = `
Array [
  Object {
    "style": Object {
      "tokens": Array [
        Object {
          "options": Array [],
          "stem": "compact-short",
        },
        Object {
          "options": Array [
            "GBP",
          ],
          "stem": "currency",
        },
      ],
      "type": 0,
    },
    "type": 2,
    "value": "0",
  },
]
`;

exports[`case: "compact-short currency/GBP" 2`] = `"{0, number, ::compact-short currency/GBP}"`;

exports[`case: "currency/CAD unit-width-narrow" 1`] = `
Array [
  Object {
    "style": Object {
      "tokens": Array [
        Object {
          "options": Array [
            "CAD",
          ],
          "stem": "currency",
        },
        Object {
          "options": Array [],
          "stem": "unit-width-narrow",
        },
      ],
      "type": 0,
    },
    "type": 2,
    "value": "0",
  },
]
`;

exports[`case: "currency/CAD unit-width-narrow" 2`] = `"{0, number, ::currency/CAD unit-width-narrow}"`;
